### ISSUE TITLE
`task`: make `workdir.output` more intuitive

### ISSUE THREAD

**Issue by casperdcl** [iterative/terraform-provider-iterative#414]
right now task `destroy` does `cp remote://input local://output` but I think we need `cp remote://output local://output`

- more intuitive
- consistent with task `apply` (`cp local://input remote://input`)

[detailed before-and-after spec](https://github.com/iterative/terraform-provider-iterative/issues/414#issuecomment-1061321093)

**Comment by 0x2b3bfa0** [comment]
(Seeing that nobody jumps it)

I like this idea and believe that it's a good compromise between the two opposed approaches we've been discussing all this time:

1. Don't impose any specific project structure and overwrite the whole working directory.
2. Require users to specify a separate path for outputs.

**Comment by 0x2b3bfa0** [comment]
Examples of increasing complexity following this proposal:

```hcl
resource "iterative_task" "example" {
  ···
  storage {
    workspace = "."
    artifacts = "./output"
  }
}
```

```hcl
resource "iterative_task" "example" {
  ···
  storage {
    workspace = "."
    artifacts = ["./output", "./models/**.hdf5"]
  }
}
```

```hcl
resource "iterative_task" "example" {
  ···
  storage {
    workspace   = "."
    artifacts   = ["./output", "./models/**.hdf5"]
    output = "./experiments/1"
  }
}
```

Don't get distracted by naming, it's subject to change.

**Comment by 0x2b3bfa0** [comment]
`workspace` — what to upload
`artifacts` — what to download
`output` (optional) — where to download `artifacts`; defaults to `workspace`

**Comment by 0x2b3bfa0** [comment]
All of this functionality can be trivially implemented with the `rclone` Go module

**Comment by DavidGOrtega** [comment]
> All of this functionality can be trivially implemented with the rclone Go module.

That's exciting. Are we talking about also of running within a go executor 😃

**Comment by 0x2b3bfa0** [comment]
No, we are talking about using `rclone` as we already do, but taking advantage of  its include/exclude functionality.

**Comment by 0x2b3bfa0** [comment]
By the way, @DavidGOrtega, you were assigned to this issue because of the discussion. What do you think about these proposals? Not about naming, but functionality.

**Comment by casperdcl** [comment]
strongly against [`art[ie]facts`](https://github.com/iterative/terraform-provider-iterative/issues/414#issuecomment-1060089232). Far too complicated and a major anti-pattern IMO (user code writes data to path X, but iff run via TPI we'll move it from X to Y). That would be for discussion in a separate issue anyway.

I'd rather just change the current behaviour of `workdir.output` for now.

**Comment by 0x2b3bfa0** [comment]
Sounds fair. [How] would we deal with https://github.com/iterative/terraform-provider-iterative/issues/306?

**Comment by 0x2b3bfa0** [comment]
> _user code writes data to path X, but iff run via TPI we'll move it from X to Y_

Directory structure would be preserved.

**Comment by casperdcl** [comment]
I... don't know what #306 is. At some point everyone agreed it was a duplicate of #307 (which is fully supported by #414 here).

However @DavidGOrtega [changed his mind with a comment I couldn't follow](https://github.com/iterative/terraform-provider-iterative/issues/306#issuecomment-980024757).

**Comment by 0x2b3bfa0** [comment]
#306 is (or should be) an error that occurs when several `rclone copy` processes try to write the same file at the same time. For example, when users try to download results from several tasks at once, saving them to the same directory,

**Comment by 0x2b3bfa0** [comment]
How does #414 (self) address this case? Can you share an example HCL block and the desired behavior?

**Comment by casperdcl** [comment]
```hcl
resource "iterative_task" "example1" {
  ···
  storage {
    input = "."
    output = "one"
  }
}

resource "iterative_task" "example2" {
  ···
  storage {
    input = "."
    output = "two"
  }
}
```

**Comment by 0x2b3bfa0** [comment]
And running `terraform destroy` would, in parallel:
* `rclone copy remote://{workdir}/one local://{input}/one`
* `rclone copy remote://{workdir}/two local://{input}/two`
...where `remote://{workdir}` is the place where files from `{input}` were uploaded.

**Comment by 0x2b3bfa0** [comment]
Users would be responsible of storing results under `{workdir}/{output}` for every task.

**Comment by casperdcl** [comment]
> Users would be responsible of storing results under {workdir}/{output} for every task.

Yes, exactly as they would intuitively do when running jobs in parallel locally (before clouds existed).

**Comment by 0x2b3bfa0** [comment]
Crystal clear. Thank you very much!

**Comment by DavidGOrtega** [comment]
@0x2b3bfa0 any stopper that you might discuss?

**Comment by 0x2b3bfa0** [comment]
Fortunately, I haven't found any major stoppers. Thanks anyway! 🙏🏼

**Comment by 0x2b3bfa0** [comment]
Still I like the idea of having several output globs, but it's more of a future–facing API choice than a real need for basic use cases.

```hcl
resource "iterative_task" "example1" {
  script = "./run.py --metrics=metrics.json --output=./output"
  storage {
    input  = "."
    outputs = ["output", "metrics.json"]
  }
}
```

Also, a bikesheddish remark: should we rename `input` to `workspace` or `workdir`?

**Comment by casperdcl** [comment]
I think `workspace` is a good idea for future compatibility with Git/DVC-workspace-awareness.

In terms of outputs being a list, does the schema support entries that are `Union[str, List[str]]`, or must we chose between list & str?

**Comment by 0x2b3bfa0** [comment]
No, it would have to be a [list](https://www.terraform.io/plugin/sdkv2/schemas/schema-types#typelist), even if users only want to specify a single directory.

```hcl
resource "iterative_task" "example1" {
  script = "./run.py --metrics=metrics.json --output=./output"
  storage {
    workspace  = "."
    outputs = ["output"]
  }
}
```

**Comment by casperdcl** [comment]
In that case I'd prefer UNIX & Windows-style path separators (`:`), i.e. "output:metrics.json:another_dir" in some future follow-up PR.

**Comment by 0x2b3bfa0** [comment]
Argh! In my opinion, this is wrong at so many levels. 🤯

**Comment by 0x2b3bfa0** [comment]
What if we ship `output: string` with this pull request and add `outputs: list` in the future? Functionally, `output` would only be an alias for `outputs = [output]`

```hcl
storage {
  workspace  = "."
  output = "output"
}
```

```hcl
storage {
  workspace  = "."
  outputs = ["one", "two", "three/**.json"]
}
```

**Comment by 0x2b3bfa0** [comment]
This kind of behavior has several precedents in official providers. For example:

> * _[assignable_scopes](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/role_definition#assignable_scopes) - (Optional) One or more assignable scopes for this Role Definition_
> * _[scope](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/role_definition#scope) - (Required) The scope at which the Role Definition applies to. It is recommended to use the first entry of the assignable_scopes._
> 
> _The value for scope is automatically included in this list if no other values supplied._
